ethereum/contracts/state-transition/StateTransition.sol

```diff
@@ -0,0 +1,212 @@
+        initData = bytes.concat(
+            IDiamondInit.initialize.selector,
+            bytes32(_chainId),
+            bytes32(uint256(uint160(address(bridgehub)))),
+            bytes32(uint256(uint160(address(this)))),
+            bytes32(uint256(stateTransitionStorage.protocolVersion)),
+            bytes32(uint256(uint160(_governor))),
+            bytes32(uint256(uint160(_governor))),
+            bytes32(stateTransitionStorage.storedBatchZero),
+            copiedData
+        );
+        Diamond.DiamondCutData memory cutData = _diamondCut;
+        cutData.initCalldata = initData;
+
+        // deploy stateTransitionChainContract
+        DiamondProxy stateTransitionChainContract = new DiamondProxy(block.chainid, cutData);
+
+        // save data
+        address stateTransitionChainAddress = address(stateTransitionChainContract);
+
+        stateTransitionStorage.stateTransitionChainContract[_chainId] = stateTransitionChainAddress;
+        stateTransitionStorage.chainNumberToContract[stateTransitionStorage.totalChains] = stateTransitionChainAddress;
+        ++stateTransitionStorage.totalChains;
+
+        IBridgehub(bridgehub).setStateTransitionChainContract(_chainId, stateTransitionChainAddress);
+
+        // set chainId in VM
+        _setChainIdUpgrade(_chainId, stateTransitionChainAddress);
+
+        emit StateTransitionNewChain(_chainId, stateTransitionChainAddress);
+    }
+
+    function setUpgradeDiamondCut(
+        Diamond.DiamondCutData calldata _cutData,
+        uint256 _protocolVersion
+    ) external onlyGovernor {
+        stateTransitionStorage.upgradeCutHash[_protocolVersion] = keccak256(abi.encode(_cutData));
+    }
+
+    function upgradeChain(
+        uint256 _chainId,
+        uint256 _protocolVersion,
+        Diamond.DiamondCutData calldata _cutData
+    ) external onlyChainGovernor(_chainId) {
+        bytes32 cutHash = keccak256(abi.encode(_cutData));
+        require(cutHash == stateTransitionStorage.upgradeCutHash[_protocolVersion], "r25");
+
+        IStateTransitionChain stateTransitionChainContract = IStateTransitionChain(
+            stateTransitionStorage.stateTransitionChainContract[_chainId]
+        );
+        stateTransitionChainContract.executeUpgrade(_cutData, _protocolVersion);
+    }
+
+    function specialUpgradeChain(
+        uint256 _chainId,
+        uint256 _protocolVersion,
+        Diamond.DiamondCutData calldata _cutData
+    ) external onlyGovernor {
+        IStateTransitionChain stateTransitionChainContract = IStateTransitionChain(
+            stateTransitionStorage.stateTransitionChainContract[_chainId]
+        );
+        stateTransitionChainContract.executeUpgrade(_cutData, _protocolVersion);
+    }
+
+    function freezeNotUpdated() external onlyGovernor {
+        uint256 protocolVersion = stateTransitionStorage.protocolVersion;
+        for (uint256 i = 0; i < stateTransitionStorage.totalChains; i = i.uncheckedInc()) {
```

Comment:
This method works in `O(stateTransitionStorage.totalChains)`. If `stateTransitionStorage.totalChains` grows too large, this method will become unfittable into Ethereum block. 

It has to be rewritten to something that can always fit into one L1 block. What's more, in my opinion it should be done in O(1) by the governor & all the state transition chains must check that they have an acceptable version in their `Executor`

Reply:
I agree with the first point, we will have to split it up. 

On the second point I disagree, because we will have to do this check when we commit/prove/execute all blocks, that seems more work than is strictly necessary.

Reply:
@kelemeno I understand your concerns about additional costs, but IMHO in case we have tens of thousands of chains to which we have to make an emergency upgrade, this might become a big security consideration. In the short term we of course will likely have an amount which can fit into 1-2 Ethereum blocks. However, with a permissionless system, we might accumulate large amounts of contracts, which will result in such mass-freeze being hard & expensive to make.